File: Data Structures & Algorithms/valid-tree/submission-1.py
class Solution:
	def validTree(self, n: int, edges: List[List[int]]) -> bool:
		if n-1 != len(edges):
			return False

		indegree = defaultdict(lambda: 0)
		graph = defaultdict(list)
		for from_node, to_node in edges:
			graph[from_node].append(to_node)
			indegree[to_node] += 1

		queue = collections.deque()
		for node in range(n):
			if indegree[node] > 1:
				return False
			if indegree[node] == 0:
				queue.append(node)
				del indegree[node]
		
		while queue:
			node = queue.popleft()
			for neighbor in graph[node]:
				indegree[neighbor] -= 1
				queue.append(neighbor)
				del indegree[neighbor]
		
		if len(indegree) > 0:
			return False
		
		return True

File: Data Structures & Algorithms/course-schedule-ii/submission-1.go
func findOrder(numCourses int, prerequisites [][]int) []int {
	indegree := make(map[int]int, numCourses)
	graph := make(map[int][]int, numCourses)
	for _, p := range prerequisites {
		graph[p[1]] = append(graph[p[1]], p[0])
		indegree[p[0]]++
	}

	var queue []int
	for course := range numCourses {
		if indegree[course] == 0 {
			queue = append(queue, course)
		}
	}

	result := make([]int, 0, numCourses)
	for len(queue) > 0 {
		course := queue[0]
		queue = queue[1:]
		result = append(result, course)
		for _, neighbor := range graph[course] {
			indegree[neighbor]--
			if indegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	if len(result) != numCourses {
		return []int{}
	}

	return result
}
